### PR TITLE
[Impeller] Support alpha in Gradients, fix uneven stops.

### PR DESCRIPTION
**Pull request by jonahwilliams**
Gives all gradients/color source contents an alpha. Additionall corrects some implementation in errors when lerping gradients

Opacity:

Screen Shot 2022-09-12 at 10 23 54 AM

Fixes Stops:

Screen Shot 2022-09-12 at 10 46 50 AM

Fixes https://github.com/flutter/flutter/issues/111313

### CODE REVIEW

File: impeller/geometry/gradient.cc [hunk 1]
```diff
@@ -59,10 +59,9 @@ std::vector<uint8_t> CreateGradientBuffer(const std::vector<Color>& colors,
     AppendColor(previous_color, &color_stop_channels);
 
     for (auto i = 1u; i < texture_size - 1; i++) {
-      auto scaled_i = i / texture_size;
+      auto scaled_i = i / (texture_size * 1.0);
```
Comment (jonahwilliams) [inline impeller/geometry/gradient.cc:62]: woops

File: impeller/geometry/gradient.cc [hunk 2]
```diff
@@ -77,16 +76,17 @@ std::vector<uint8_t> CreateGradientBuffer(const std::vector<Color>& colors,
...
         AppendColor(mixed_color, &color_stop_channels);
       } else {
-        // We've slightly overshot the next stop. In theory this only happens if
-        // we have scaled our texture such that not every stop gets their own
-        // index. For now I am simply ignoring the inbetween colors. Currently
-        // this requires a gradient with either an absurd number of textures
-        // or very small stops.
-        AppendColor(next_color, &color_stop_channels);
-
+        // We've slightly overshot the previous stop.
```
Comment (jonahwilliams) [inline impeller/geometry/gradient.cc:79]: This isn't the case I though it was. If we've passed the previous stop, then we essentially just treat the current stop as the previous stop and run the same code as above.

**Review by bdero — APPROVED** [review]
LGTM! Should we land this for the other color sources too? (Could be done in followup(s))

File: impeller/entity/shaders/linear_gradient_fill.frag [hunk 3]
```diff
@@ -30,5 +31,5 @@ void main() {
     gradient_info.texture_sampler_y_coord_scale,
     gradient_info.tile_mode,
     gradient_info.tile_mode);
-  frag_color = vec4(frag_color.xyz * frag_color.a, frag_color.a);
+  frag_color = vec4(frag_color.xyz * frag_color.a, frag_color.a) * gradient_info.alpha;
```
Comment (dnfield) [inline impeller/entity/shaders/linear_gradient_fill.frag:34]: I left this comment on the other PR, but this still makes me wonder if `IPSampleWithTileMode` shouldn't really just be `IPSampleWithTileModeAndAlpha` so we don't forget to do this somewhere - e.g. if and when we implement two point conical gradients.

**Review by dnfield — APPROVED** [review]

File: impeller/entity/shaders/linear_gradient_fill.frag (hunk 3 above)
Comment (bdero) [reply impeller/entity/shaders/linear_gradient_fill.frag:34]: One idea for making this a bit more semantic is to add an `IPPremultiply` utility to the shader library and wrap the sampling function in it. This way new contributors who don't know why we're doing the funky alpha multiplies can get a bit more context when first encountering it.

**Pull request merged**